src/Stratis.Bitcoin/Connection/PeerLifetime.cs

```diff
@@ -0,0 +1,145 @@
+﻿using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.Utilities;
+using System;
+using System.Collections.Concurrent;
+using System.Net;
+
+namespace Stratis.Bitcoin.Connection
+{
+    /// <summary>
+    /// A class that will manage the lifetime of connected peers and be in charge of banning malicious peers.
+    /// </summary>
+    /// <remarks>
+    /// By default peers are banned for 24 hours, this value can change using configuration (-bantime=[seconds]).
```

Comment:
Initially I did have 7 days, but then decided to do like core (which has it at 24h).
Interesting idea of keeping history but is that really necessarily, I would increase the ban time rather then keep history.

Rate limiting in what way? a node that was previously marked is now allowed only a certain amount of blocks?

Reply:
i don't know if it is necessary, but having a one bantime to fit them all seems weird to me. but if we make the design little more flexible in this, we will be able to tweak the protection per every use case. the trade off is little more time invested at the beginning. risk is that we then find it to be an overkill. but in general, if there is a possibility of a node being somehow faulty, but not malicious, we should be able to ban it for short time instead of long time. if we go with 1 week for everything, it might be little too much. tough decision, as always :) i don't mind little more complexity for better flexibility

as for rate limiting - there are many different message types on p2p network and each has different meaning and usage and for each of those types you could have different number of allowed frequency. for example, if someone sends you ping every second, i'd say that is not OK, rather ban that guy after certain threshold. same for every message type. however, i don't know if something like this is not there already (but probably not as we don't have any banning implemented).

Reply:
I would say yes if we find a need to ban for for different reasons then we can add that functionality at that point.

As for rate limiting yes I like the idea, this can be a new PR on the PeerLifetimeBehaviour that will listen to all messages and increment a counter, I like that idea see #722